crdt.go

```diff
@@ -0,0 +1,531 @@
+// batch if they were added.
+func (store *Datastore) updateDeltaWithRemove(key string, newDelta *pb.Delta) {
+	store.curDeltaMux.Lock()
+	defer store.curDeltaMux.Unlock()
+	elems := make([]*pb.Element, 0)
+	for _, e := range store.curDelta.GetElements() {
+		if e.GetKey() != key {
+			elems = append(elems, e)
+		}
+	}
+	store.curDelta = &pb.Delta{
+		Elements:   elems,
+		Tombstones: store.curDelta.GetTombstones(),
+		Priority:   store.curDelta.GetPriority(),
+	}
+	store.curDelta = deltaMerge(store.curDelta, newDelta)
+}
+
+func (store *Datastore) updateDelta(newDelta *pb.Delta) {
+	store.curDeltaMux.Lock()
+	defer store.curDeltaMux.Unlock()
+	store.curDelta = deltaMerge(store.curDelta, newDelta)
+}
+
+func (store *Datastore) publishDelta() error {
+	store.curDeltaMux.Lock()
+	defer store.curDeltaMux.Unlock()
+	err := store.publish(store.curDelta)
+	if err != nil {
+		return err
+	}
+	store.curDelta = nil
+	return nil
+}
+
+func (store *Datastore) putBlock(heads []cid.Cid, height uint64, delta *pb.Delta) (cid.Cid, error) {
+	if delta != nil {
+		delta.Priority = height
+	}
+	node, err := makeNode(delta, heads)
+	if err != nil {
+		return cid.Undef, errors.Wrap(err, "error creating new block")
+	}
+	err = store.dags.Add(store.ctx, node)
+	if err != nil {
+		return cid.Undef, errors.Wrapf(err, "error writing new block %s", node.Cid())
+	}
+	return node.Cid(), nil
+}
+
+func (store *Datastore) publish(delta *pb.Delta) error {
+	c, err := store.addDAGNode(delta)
+	if err != nil {
+		return err
+	}
+	return store.broadcast(c)
+}
+
+func (store *Datastore) addDAGNode(delta *pb.Delta) (cid.Cid, error) {
+	heads, height, err := store.heads.List()
+	if err != nil {
+		return cid.Undef, errors.Wrap(err, "error listing heads")
+	}
+	height = height + 1
+
+	delta.Priority = height
+
+	// for _, e := range delta.GetElements() {
+	// 	e.Value = append(e.GetValue(), []byte(fmt.Sprintf(" height: %d", height))...)
+	// }
+
+	c, err := store.putBlock(heads, height, delta)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	// Process new block. This makes that every operation applied
+	// to this store takes effect (delta is merged) before
+	// returning. Optimizations possible by not fetching the block
+	// again from the DAG and passing it directly.
+	store.logger.Debugf("processing generated block %s", c)
+	err = store.walkBranch(c, c, 1)
+	if err != nil {
+		return cid.Undef, errors.Wrap(err, "error processing new block")
+	}
+
+	return c, nil
+}
+
+func (store *Datastore) broadcast(c cid.Cid) error {
+	if store.broadcaster == nil { // offline
+		return nil
+	}
+	store.logger.Debugf("broadcasting new block %s", c)
+	err := store.broadcaster.Broadcast(c.Bytes())
+	if err != nil {
+		return errors.Wrapf(err, "error broadcasting new block %s", c)
+	}
+	return nil
+}
+
+// IsThreadSafe declares that this datastore implementation is thread-safe.
+func (store *Datastore) IsThreadSafe() {}
```

Comment:
Fill this or add a `TODO` ?

Reply:
There is nothing todo. That's just how a datastore implements the ThreadSafe interface for datastores.